### PR TITLE
Delete iot device with transaction

### PR DESCRIPTION
**Pull request by AramAlsabti**

### CODE REVIEW

File: src/services/device-management/iot-device.service.ts [hunk 1]
```diff
@@ -467,15 +468,54 @@ export class IoTDeviceService {
...
-                `Deleting LoRaWANDevice ${lorawanDevice.id} / ${lorawanDevice.deviceEUI} in Chirpstack ...`
-            );
-            await this.chirpstackDeviceService.deleteDevice(lorawanDevice.deviceEUI);
+        let deleteResult: DeleteResult = {
+            raw: null,
+            affected: 0
         }
 
-        return this.iotDeviceRepository.delete(device.id);
+        // If operations on chirpstack fail, rollback any mutations on the database
```
Comment (GufCab) [inline src/services/device-management/iot-device.service.ts:476]: "Run these operations inside a Typeorm transaction to make sure if operations on chirpstack fail, rollback any changes on the database to ensure consistency between devices in the OS2iot database and the Chirpstack database"

File: src/services/device-management/iot-device.service.ts [hunk 2]
```diff
@@ -467,15 +468,54 @@ export class IoTDeviceService {
...
+            // Find and delete m-n relations with the device first
+            const iotDeviceRepository = transactionManager.getRepository(IoTDevice);
+            const multicastRepository = transactionManager.getRepository(Multicast);
+            // Get all multicasts. For some reason, filtering by device id won't return multicasts with all devices
+            const _multicasts = await multicastRepository
+                .createQueryBuilder("multicast")
+                .innerJoinAndSelect("multicast.iotDevices", "iot_device")
+                .getMany();
+
+            // Filter multicasts without the device out to avoid updating them
```
Comment (GufCab) [inline src/services/device-management/iot-device.service.ts:487]: "Take only the multicasts that has references to the device to be deleted"

File: src/services/device-management/iot-device.service.ts [hunk 3]
```diff
@@ -467,15 +468,54 @@ export class IoTDeviceService {
...
         }
 
-        return this.iotDeviceRepository.delete(device.id);
+        // If operations on chirpstack fail, rollback any mutations on the database
+        await getManager().transaction((async transactionManager => {
+            // Find and delete m-n relations with the device first
+            const iotDeviceRepository = transactionManager.getRepository(IoTDevice);
+            const multicastRepository = transactionManager.getRepository(Multicast);
+            // Get all multicasts. For some reason, filtering by device id won't return multicasts with all devices
+            const _multicasts = await multicastRepository
```
Comment (GufCab) [inline src/services/device-management/iot-device.service.ts:482]: Refactor: Move all multicast deletion to its own method.

File: src/services/device-management/iot-device.service.ts [hunk 4]
```diff
@@ -467,15 +468,54 @@ export class IoTDeviceService {
...
-            );
-            await this.chirpstackDeviceService.deleteDevice(lorawanDevice.deviceEUI);
+        let deleteResult: DeleteResult = {
+            raw: null,
+            affected: 0
         }
 
-        return this.iotDeviceRepository.delete(device.id);
+        // If operations on chirpstack fail, rollback any mutations on the database
+        await getManager().transaction((async transactionManager => {
```
Comment (GufCab) [inline src/services/device-management/iot-device.service.ts:477]: Are we also handling the issue with the datatarget? As far as I can see we are only handling the case with multicast.

**Review by GufCab — CHANGES_REQUESTED** [review]

File: src/services/device-management/iot-device.service.ts (hunk 4 above)
Comment (AramAlsabti) [reply src/services/device-management/iot-device.service.ts:477]: Right before deleting the device, we set the data targets, `connections`, to `[]` and save it.

**Review by GufCab — APPROVED** [review]

**Pull request merged**